src/librustc_middle/ty/sty.rs

```diff
@@ -1623,16 +1659,19 @@ impl RegionKind {
                 flags = flags | TypeFlags::HAS_FREE_LOCAL_REGIONS;
                 flags = flags | TypeFlags::HAS_RE_INFER;
                 flags = flags | TypeFlags::KEEP_IN_LOCAL_TCX;
+                flags = flags | TypeFlags::STILL_FURTHER_SPECIALIZABLE;
```

Comment:
Hmm, I'm not sure on lifetimes affecting specialization, but it's probably good to keep this in, just in case there are edge cases around higher-ranked `fn` pointer types.

Reply:
> I'm not sure on lifetimes affecting specialization

That is the whole reason specialization is unsound, right?

Reply:
I'm not worried about the unsound cases but rather things like `impl Foo for for<'a> fn(&'a ())` vs `impl<'a> Foo for fn(&'a ())` which AFAIK are both sound for specialization and still allowed as non-overlapping.